dist: introduce scylla-tune-sched.service to tune kernel scheduler

syuu1228 opened the pull request:
On /usr/lib/sysctl.d/99-scylla-sched.conf, we have some sysctl settings to tune the scheduler for lower latency.
This is mostly to prevent softirq threads processing tcp and reactor threads from injecting latency into each other.
However, these parameters are moved to debugfs from linux-5.13+, so we lost scheduler tuneing on recent kernels.

To tune recent kernel, let's add a new service to configure debugfs with sysctl like configuration file.
The service named scylla-debugfs.service, and configuration file is placed /etc/scylla.d/debugfs.yaml.

The service will unconditionally enables when installed, but does nothing on older kernels since it will ignore File not found error.

Fixes #16077

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - Build
:white_check_mark: - Sanity Tests
:white_check_mark: - Unit Tests
#### Build Details:
- Build URL: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4820/
- Duration: 2 hr 25 min
- Builder: spider1.cloudius-systems.com

tchaikov commented on dist/common/systemd/scylla-debugfs.service:
```diff
@@ -0,0 +1,10 @@
+[Unit]
+Description=Scylla debugfs configrator
+After=local-fs.target
```
instead of starting this service after `local-fs` is fully ready, i'd suggest add procfs and debugfs directories to `RequiresMountsFor=`, see https://www.freedesktop.org/software/systemd/man/latest/systemd.unit.html#RequiresMountsFor=

tchaikov commented on dist/common/systemd/scylla-debugfs.service:
```diff
@@ -0,0 +1,10 @@
+[Unit]
+Description=Scylla debugfs configrator
```
typo.

tchaikov commented on dist/common/systemd/scylla-debugfs.service:
```diff
@@ -0,0 +1,10 @@
+ExecStart=/usr/sbin/scylla_debugfs
+
+[Install]
+WantedBy=local-fs.target
```
strictly speaking, i don't think local-fs should depend on this service. what should depend on it is, IMO, `scylla-server` or `multi-user`. what do you think?

syuu1228 commented on dist/common/scylla.d/debugfs.yaml:
```diff
@@ -0,0 +1,11 @@
+sched.tunable_scaling: 0
+
+# Preempt sooner
+sched.min_granularity_ns: 500000
```
> might want add
> 
> ```yaml
> sched.base_slice_ns: 0
> ```
> 
> as an alternative of `sched.tunable_scaling` . see [torvalds/linux@e4ec331](https://github.com/torvalds/linux/commit/e4ec3318a17f5dcf11bc23b2d2c1da4c1c5bb507). the rename reflect is usage in EEVDF.

Seems like it's alternative of sched.min_granularity_ns, not sched.tunable_scaling, no?

tchaikov replied:
ahh, right. sorry for the copy-pasta.

syuu1228 commented:
@tchaikov Pushed new version with following changes:
 - renamed scylla_debugfs to scylla_tune_sched
     - support to tune both sysctl and debugfs
     - drop configuration file and move parameters to script itself, since the script just for scheduler tuning, not generic script for debugfs
     - drop 99-scylla-sched.conf since now we support sysctl
 - drop encoding declaration
 - add sched.base_slice_ns to support EEVDF
 - add comments on sched.wakeup_granularity_ns and sched.latency_ns to describe these parameters are dropped on linux-6.6+
 - Added RequiresMountsFor= on scylla-tune-sched.service
 - Replaced local-fs.target to multi-user.target on scylla-tune-sched.service

tchaikov commented on dist/common/scripts/scylla_tune_sched:
```diff
@@ -0,0 +1,53 @@
+
+import os
+import sys
+import yaml
```
this import is not used.

syuu1228 replied:
Dropped `import yaml`.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - Build
:white_check_mark: - Sanity Tests
:white_check_mark: - Unit Tests
#### Build Details:
- Build URL: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4918/
- Duration: 2 hr 9 min
- Builder: spider5.cloudius-systems.com

avikivity commented on dist/common/scripts/scylla_tune_sched:
```diff
@@ -0,0 +1,53 @@
+    'sched.min_granularity_ns': 500000,
+
+    # Preempt sooner  (For EEVDF, only available on >=linux-6.6)
+    'sched.base_slice_ns': 500000,
```
@xemul we should examine EEVDF and see if we can apply any lessons to the Seastar scheduler

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5010/)
:white_check_mark: - [Sanity Tests](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/6369/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5010/)
#### Build Details:
- Duration: 2 hr 11 min
- Builder: spider2.cloudius-systems.com

tchaikov commented on dist/redhat/scylla.spec:
```diff
@@ -84,16 +84,18 @@ getent group scylla || /usr/sbin/groupadd scylla 2> /dev/null || :
     /usr/bin/systemctl preset scylla-server.service ||:
+    /usr/bin/systemctl preset scylla-tune-sched.service ||:
 fi
+/opt/scylladb/scripts/scylla_post_install.sh
```
@syuu1228 hi Takuya, could you reconsider this suggestion?

avikivity replied:
Ah, I see the existing code is in kernel-conf. So it needs to be kept there.

avikivity commented on dist/common/scripts/scylla_tune_sched:
```diff
@@ -0,0 +1,52 @@
+    if not os.path.isfile(path):
+        return False
+    with open(path, 'w') as f:
+        f.write(str(value))
```
This could fail (example: in a container, without elevated permissions). So need to catch the exception and return False.

avikivity commented:
> @syuu1228 - we should not be working with 5.13. Only 5.15 (LTS) kernel. Is the bug still relevant there?

We can't control the installed kernel, better to be prepared for any.

avikivity commented:
> > > @syuu1228 - we should not be working with 5.13. Only 5.15 (LTS) kernel. Is the bug still relevant there?
> > 
> > 
> > We can't control the installed kernel, better to be prepared for any.
> 
> We can, for example in the setup script. There are at least 2 bugs that I recall in 5.13 (md0 not reporting disk stats) and 5.19 (the AWS ena interface issue) that we should stick to 5.15 LTS

It just leads to "can't install" bugs. Some corporate environments don't give any platform choice. It will also break upgrades.
